socs/agents/hwp_pmx/agent.py

```diff
@@ -0,0 +1,277 @@
+                    'Could not use external voltage because {} is already running'.format(self.lock.job))
+                return False, 'Could not acquire lock'
+            self.dev.use_external_voltage()
+        return True, 'Set PMX Kikusui to PID control'
+
+    def ign_ext(self, session, params):
+        """ign_ext()
+        **Task** - Set the PMX Kiksui to ignore external voltage control. Doing so
+        disables the PID and switches to direct control.
+        """
+        with self.lock.acquire_timeout(3, job='ign_ext') as acquired:
+            if not acquired:
+                self.log.warn(
+                    'Could not ignore external voltage because {} is already running'.format(self.lock.job))
+                return False, 'Could not acquire lock'
+            self.dev.ign_external_voltage()
+        return True, 'Set PMX Kikusui to direct control'
+
+    def start_acq(self, session, params):
+        """acq
+        Method to start data acquisition process.
+        The most recent data collected is stored in session.data in the
+        structure::
+
+            >>> response.session['data']
+            {'curr': 0,
+             'volt': 0,
+             'last_updated': 1649085992.719602}
+        """
+        sleep_time = 1 / self.f_sample - 0.01
+
+        with self.lock.acquire_timeout(0, job='acq') as acquired:
+            if not acquired:
+                self.log.warn("Could not start acq because {} is already running"
+                              .format(self.lock.job))
+                return False, "Could not acquire lock."
+
+            session.set_status('running')
+            self.take_data = True
+
+            last_release = time.time()
+            while self.take_data:
+                current_time = time.time()
+                if current_time - last_release > 1.:
+                    last_release = time.time()
+                    if not self.lock.release_and_acquire(timeout=10):
+                        self.log.warn(f"Failed to re-acquire lock, currently held by {self.lock.job}.")
+                        continue
+
+                data = {
+                    'timestamp': current_time,
+                    'block_name': 'Rotation',
+                    'data': {}
+                }
+                msg, curr = self.dev.meas_current()
+                data['data']['current'] = curr
+                msg, volt = self.dev.meas_voltage()
+                data['data']['voltage'] = volt
+
+                self.agent.publish_to_feed('Rotation', data)
+                session.data = {'curr': curr,
+                                'volt': volt,
+                                'last_updated': current_time}
+                time.sleep(sleep_time)
+            self.agent.feeds['Rotation'].flush_buffer()
+        return True, 'Acquisition exited cleanly.'
+
+    def stop_acq(self, session, params=None):
+        """
+        Stop acq process.
+        """
+        self.dev.close()
```

Comment:
Thanks for pointing out.
I added close method here https://github.com/simonsobs/socs/pull/421/commits/21a0a080e35c616c888e865d1faea2eae496e1df.

Reply:
Ah, great. However, now if the acquisition process is stopped it will close the connection to the device and it will be required that `init_connection` be run before `acq` can run again. Is this the desired behavior?

I think `self.dev.close()` shouldn't be called in the `_stop_acq()` function. That way the connection stays open and doesn't need to be reinitialized. This is the behavior in other Agents.

Reply:
I see. Thanks.
I simply removed `self.dev.close()` from `_stop_acq()` in the same way as other agents. https://github.com/simonsobs/socs/pull/421/commits/774008b49936eb78447fc0ae5e31e051dd0a7aa9